aten/src/ATen/native/sparse/SparseCsrTensor.cpp

```diff
@@ -103,38 +103,49 @@ void _validate_sparse_compressed_tensor_args_worker(const Tensor& compressed_ind
       "number of dimensions of ", compressed_indices_name, " and ", plain_indices_name, " must be the same but got ",
       compressed_indices.dim(), " and ", plain_indices.dim(), ", respectively");
 
-  AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(
+  int block_ndim = AT_DISPATCH_PLAIN_SPARSE_COMPRESSED_LAYOUTS(
       layout, "validate_sparse_compressed_tensor_args",
       [&] {
         TORCH_CHECK(
-                    compressed_indices.dim() == values.dim(),
-                    "number of dimensions of indices and values must be the same but got ",
-                    compressed_indices.dim(), " and ", values.dim(), ", respectively");
+                    compressed_indices.dim() <= values.dim(),
+                    "number of dimensions of indices (=", compressed_indices.dim(),
+                    ") must be equal or less than the number of dimensions of values (=", values.dim(), ")");
+        return 0;
       },
       [&] {
         TORCH_CHECK(
-                    compressed_indices.dim() + 2 == values.dim(),
-                    "number of dimensions of indices must be two less than the number of dimensions of the values but got ",
-                    compressed_indices.dim(), " + 2 not equal to ", values.dim());
+                    compressed_indices.dim() + 2 <= values.dim(),
+                    "number of dimensions of indices (=", compressed_indices.dim(),
+                    ") plus two must be equal or less than the number of dimensions of values (=", values.dim(), ")");
+        return 2;
       });
+  int dense_ndim = values.dim() - compressed_indices.dim() - block_ndim;
+  TORCH_CHECK(dense_ndim == 0, "non-zero dense dimensions (=", dense_ndim, ") is not supported for ", layout, " layout");
+
+  int batch_ndim = size.size() - 2 - dense_ndim;
+  TORCH_INTERNAL_ASSERT(block_ndim >= 0 && dense_ndim >=0 && batch_ndim >= 0);
 
   TORCH_CHECK(
-      static_cast<size_t>(compressed_indices.dim()) == size.size() - 1,
-      "number of dimensions of indices must be one less than the number of dimensions of the provided size but got ",
-      compressed_indices.dim(), " not equal to ", size.size(), " - 1");
+      static_cast<size_t>(compressed_indices.dim()) == size.size() - 1 - dense_ndim,
+      "number of dimensions of indices must be one less than the number of dimensions of the provided size",
+      " (minus the number of dense dimensions) but got ",
+      compressed_indices.dim(), " not equal to ", size.size(), " - 1 - ", dense_ndim);
+
+  DimVector blocksize{
```

Comment:
Part of me does wonder whether it's more "sanitary" so to speak to separate out this validation into two separate functions. Essentially what this shows, and which is true, is that we can think of a regular CSR Tensor as a BSR Tensor with 1 by 1 blocks. Maybe eventually SparseTensorImpl.h will be able to do it all.